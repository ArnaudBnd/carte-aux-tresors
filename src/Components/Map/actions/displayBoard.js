import React, { Component }  from 'react'

/**
 * Affichage de la map
 */
export function DisplayBoard({ players, world }) {
  const button = document.getElementById('buttonPlayGame')

  if(world.length !== 0) {
    if(button.style.display === 'none') {
      button.style.display = 'block'
    }

    if(players.length > 0) {
      return world.map((elmt, i) => {
        return (
          <tr key={i}
              styles={{height: '50px'}}>
              {elmt.map((elmt1, j) => 
                <td key={j}
                  styles={{height: '50px'}}>
                  {elmt1.tresor!=0?'T('+elmt1.tresor+')':elmt1.type}
                </td>
              )}
          </tr>
        )
      })
    }
  }
}