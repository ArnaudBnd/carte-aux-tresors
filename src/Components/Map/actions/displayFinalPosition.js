import React, { Component }  from 'react'
  
/**
 * Affichage du fichier de sorti dans la console
 */
export function DisplayFinalPosition({ players, world }) {
  let coordonnesFinal = []

  players.forEach((player, index) => {
    coordonnesFinal.push('A' + ' - ' + player.name + ' - ' + player.x + ' - ' + player.y + ' - ' + player.orientation + ' - ' + player.tresors + '\n')
  })

  return (
    <div style={{textAlign: 'center'}}>
      <h1>
        RÃ©sultat:
      </h1>
      <h4 style={{whiteSpace: 'pre'}}>
        {coordonnesFinal}
      </h4>
    </div>
  )
}