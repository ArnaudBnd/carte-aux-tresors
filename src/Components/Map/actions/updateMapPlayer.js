export function update_map_player(worldMap, player){
  let coordonnee={x:0, y:0}
  let err = ""

  switch (player.orientation) {
    case 'N':
      coordonnee={y:-1, x:0}
      break
    case 'E':
      coordonnee={y:0, x:1}
      break
    case 'S':
      coordonnee={y:1, x:0}
      break
    case 'O':
      coordonnee={y:0, x:-1}
      break
  }

  try {
    if (player.y+coordonnee.y < 0 || player.x+coordonnee.x < 0 ||
        player.y+coordonnee.y >= worldMap.length || player.x + coordonnee.x >= worldMap[player.y].length) {
      throw 'player sorti de la map'
    } else if (worldMap[player.y+coordonnee.y][player.x+coordonnee.x].type === "A" ) {
      throw 'player met another player'
    } else if (worldMap[player.y+coordonnee.y][player.x+coordonnee.x].type === "0" ) {
      if(worldMap[player.y+coordonnee.y][player.x+coordonnee.x].tresor !== 0){
        player.tresors += 1
        worldMap[player.y+coordonnee.y][player.x+coordonnee.x].tresor += -1
      }
      worldMap[player.y][player.x].type = "0"
      player.y += coordonnee.y
      player.x += coordonnee.x
      worldMap[player.y][player.x].type = "A"
    } else if (worldMap[player.y+coordonnee.y][player.x+coordonnee.x].type == "M" ) {
      throw 'player met a mountain'
    }
  } catch (e) {
    err = e
    console.log(e)
  }

  return {"worldMap": worldMap, "player": player, "err": err}
}
