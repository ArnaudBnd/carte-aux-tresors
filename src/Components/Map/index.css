table {
    border-collapse: collapse;
    border: 2px solid rgb(200, 200, 200);
    letter-spacing: 1px;
    font-family: sans-serif;
    font-size: .8rem;
    text-align: center;
    margin: 0 auto;
}
td {
    border: 1px solid rgb(190, 190, 190);
    padding: 10px;
    text-align: center;
}