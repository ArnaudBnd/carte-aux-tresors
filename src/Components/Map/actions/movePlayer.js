import { update_map_player } from './updateMapPlayer'
  
export function move(movement, worldMap, player) {
  switch (movement) {
    case 'A':
      let update = update_map_player(worldMap,player)
      worldMap = update.worldMap
      player = update.player
      break
    case 'G':
      switch (player.orientation) {
        case 'N':
          player.orientation = 'O'
          break
        case 'S':
          player.orientation = 'E'
          break
        case 'O':
          player.orientation = 'S'
          break
        case 'E':
          player.orientation = 'N'
          break
      }
      break
    case 'D':
    console.log('player', player)
      switch (player.orientation) {
        case 'N':
          player.orientation = 'E'
          break
        case 'S':
          player.orientation = 'O'
          break
        case 'O':
          player.orientation = 'N'
          break
        case 'E':
          player.orientation = 'S'
          break
      }
      break
  }
  return player
}