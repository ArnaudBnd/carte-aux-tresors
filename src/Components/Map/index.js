import React, { useEffect, useState } from 'react'
import { move } from './actions/movePlayer'
import { DisplayBoard } from './actions/displayBoard'
import { DisplayFinalPosition } from './actions/displayFinalPosition'
import './index.css'

export function MapInit(props) {
  const { coordonnes } = props
  const [state, setState] = useState({
    world: [],
    mapH: '',
    mapW: '',
    players: [],
    play: false,
  })
  const { play, players, world } = state

  useEffect(() => {
    if(coordonnes.length > 0) {
      initDataBoard(coordonnes)
    }
  }, [coordonnes])

  /**
   * Initialisation des coordonnés du tableau 
   * @param coordonnes
   */
  function initDataBoard(coordonnes) {
    const world = []
    let mapH = ''
    let mapW = ''
    let players = []
    let tresors = []

    coordonnes.map(element => {
      const elmt = element[0].split('-')
      if(elmt[0].includes('C')) {
        mapH = elmt[2]
        mapW = elmt[1]

        for (var i = 0; i < elmt[2]; i++) {
          let array = []
          for (var j = 0; j < elmt[1]; j++) {
            array.push({type : '0', tresor : 0})
          }
          world.push(array)
        }
      } else if(elmt[0].includes('A')){
        world[elmt[3]][elmt[2]] = {type : 'A', tresor : 0}

        players.push({
          x: parseInt(elmt[2]),
          y: parseInt(elmt[3]),
          tresors: 0,
          orientation: elmt[4],
          movements: elmt[5].split(''),
          name: elmt[1]
        })
      } else if(elmt[0].includes('T')){
        tresors.push({
          x: parseInt(elmt[1]),
          y: parseInt(elmt[2]),
          nbr: parseInt(elmt[3])
        })
        world[elmt[2]][elmt[1]] = {type : '0', tresor : parseInt(elmt[3])}
      } else {
        world[elmt[2]][elmt[1]] = {type : elmt[0], tresor : 0}
      }
    })

    setState({...state,
      world,
      mapH,
      mapW,
      players,
      tresors
    })
  }

  /**
   * Jeu de(s) l'aventurier(s) en fonction des déplacements récupérés
   */
  function playMovement() {
    const { players, world, tresors } = state
    let no_more_movement = false
    let i = 0

    while (no_more_movement == false) {
      no_more_movement = true

      players.forEach((player, index) => {
        if(players[index].movements[i] !== undefined) {
          players[index] = move(players[index].movements[i], world, players[index], tresors)
          if(players[index].movements[i+1] !== undefined) {
            no_more_movement = false
          }
        }
      })
      i++
    }
    setState({...state, play: true})
  }

  return (
    <div>
      <table>
        <tbody>
          {world.length && players.length > 0 ? <DisplayBoard world={world} players={players} /> : null} 
        </tbody>
      </table>
      <br />
      { play ? 
          <DisplayFinalPosition world={world} players={players} /> : 
          <button 
            id="buttonPlayGame"
            style={{ display: 'none', margin: '0 auto' }}
            onClick={() => playMovement()}>
            PLAY
      </button>}
    </div>
  )
}