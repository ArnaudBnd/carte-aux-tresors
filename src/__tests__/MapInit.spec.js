import React from 'react'
import { MapInit } from '../Components/Map/index'
import { DisplayFinalPosition } from '../Components/Map/actions/displayFinalPosition'
import Enzyme from 'enzyme'
import { configure, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() });

describe("MapInit ", () => {
  let wrapper
  const setState = jest.fn()
  const useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation((init) => [init, setState])

 const coordonnes = [
    ["C​-3-4"],
    ["M​-1-0"],
    ["M​-2-1"],
    ["T​-0-3-2"],
    ["T​-1-3-3"],
    ["A​-LARA-1-1-S-AADADAGGA"],
    ["A​-NABIL-1-2-E-A"]
  ]

  beforeEach(() => {
    wrapper = shallow(<MapInit coordonnes={coordonnes} />)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('tests if component renders correctly ', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('has the initial state count of zero', () => {
    wrapper.find('button').props().onClick()
    expect(setState).not.toBe({
      world: [],
      mapH: '',
      mapW: '',
      players: [],
      play: false,
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders not null', () => {
   expect(wrapper).not.toBeNull()
  }) 

  it('contains a button PLAY', () => {
    const button = <button> PLAY </button>
    expect(wrapper.containsMatchingElement(button)).toBe(true)
  })  

  it('calls playMovement() when button PLAY is clicked', () => {
    wrapper.find('button').simulate('onClick')
  })
})