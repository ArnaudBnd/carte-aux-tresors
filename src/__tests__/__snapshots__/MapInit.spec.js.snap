// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapInit  has the initial state count of zero 1`] = `
<div>
  <table>
    <tbody />
  </table>
  <br />
  <DisplayFinalPosition
    players={Array []}
    world={Array []}
  />
</div>
`;

exports[`MapInit  tests if component renders correctly  1`] = `
<div>
  <table>
    <tbody />
  </table>
  <br />
  <button
    id="buttonPlayGame"
    onClick={[Function]}
    style={
      Object {
        "display": "none",
        "margin": "0 auto",
      }
    }
  >
    PLAY
  </button>
</div>
`;
